FE 이슈 작성 화면 image 업로드 구현 / 사이드바 select menu의 item 선택 기능 구현(& 체크 표시)

caribouuuu opened the pull request:
## 구현 내용

<img width="910" alt="스크린샷 2020-11-11 오후 11 16 29" src="https://user-images.githubusercontent.com/57661699/98822293-03978500-2474-11eb-87dc-1759915c30b2.png">
<img width="376" alt="스크린샷 2020-11-11 오후 11 16 41" src="https://user-images.githubusercontent.com/57661699/98822307-072b0c00-2474-11eb-9e0b-54fdccfd5c48.png">
<img width="384" alt="스크린샷 2020-11-11 오후 11 16 50" src="https://user-images.githubusercontent.com/57661699/98822312-07c3a280-2474-11eb-8767-8af7e93ba360.png">


###  select menu 의 item 선택 기능
- IssueFormContainer.js : assingees, labels, milestone 상태 관리
  - 각 SidebarFormContainer에 관련 항목(`setAssignees`, `setLabels`, `setMilestone`)들을 전달
- SidebarFormContainer.js : selectedItems, setSelectedItems를 전달 받아 각각에 맞는 onItemClick 이벤트 구현
  - Assignees, Labels : 배열에 추가/삭제
  - Milestone : 배열에 한 아이템만 저장/삭제
  - SelectMenuContainer에 onItemClick 함수 전달

### selected item 컴포넌트 구현
- SelectedUserItem, SelectedLabelItem, SelectedMilestoneItem

### select menu의 item 선택 시 체크 아이콘 표시
- user item, label item은 선택 시 체크 아이콘이 표시됨
  - milestone은 미구현

### 로그인/로그아웃 판단 처리
- 로그인 상태 : issue form 안의 user profile을 로그인 유저 정보로 렌더링
- 로그아웃 상태 : home으로 리다이렉트

### 새로운 issue 등록 - api 연동
- IssueFormContainer의 onSubmit 에서 issue 데이터를 다듬어 **POST /issues** 요청
  - success true인 경우 새로 생성된 id 값을 통해 생성된 이슈 상세 화면으로 이동

### imageUploadHandler 구현
- imgur로 이미지 업로드 api 요청
- link(image url) 반환
- IssueFormContainer에서 link 값을 전달받아 이미지 태그를 만들어 issue description 내용에 append


## Notes
- frontend 디렉토리에서 .env 사용을 어떻게 해야할지 몰라 일단 `/frontend/src/config/config.js` 파일로 설정 값을 저장했습니다
  - `.gitignore`에 config.js 등록
  - config.js 파일에 들어있는 정보는 IMGUR_CLIENT_ID 값입니다
- 이미지 업로드 시 multiple 파일 업로드 미구현 (구현 예정)
-  이미지 업로드 시 `'![Uploading 파일명...]()'`을 먼저 출력한 후 업로드 성공 시 완성된 이미지 태그를 replace 하는 기능은 미구현 
  - setState(setIssue)를 연속 호출할 시 생기는 문제로 인해 구현하다가 막혔습니다..
- milestone item 선택 시 체크 표시 구현은 로직상의 오류로 아직 미완성입니다.....

caribouuuu committed:
feat: select menu 컴포넌트에 overlay 추가

gear 버튼을 클릭하지 않고 select menu 바깥 영역 클릭 시 창이 닫히도록 변경
- overlay 추가하여 onOverlayClick에 open 상태 제어 이벤트 전달

caribouuuu committed:
refactor: issue form 구조 변경

- /frontend/src/pages/CreateIssue.js : IssueFormContainer 사용
- /frontend/src/containers/issue-form/IssueFormContainer.js :
  - IssueMainForm
  - Sidebar : SidebarFormContainer(Assignees, Labels, Milestone) 포함

caribouuuu committed:
feat: select menu(issue sidebar)의 item 선택 기능 구현

- IssueFormContainer.js : assingees, labels, milestone 상태 관리
  - 각 SidebarFormContainer에 useState 관련 항목들을 전달
- SidebarFormContainer.js : selectedItems, setSelectedItems를 전달받아 각각에 맞는 onItemClick 이벤트 구현
  - Assignees, Labels : 배열에 추가/삭제
  - Milestone : 배열에 한 아이템만 저장/삭제
  - SelectMenuContainer에 onItemClick 함수 전달

caribouuuu committed:
feat: SelectedUserItem 컴포넌트 추가

- SidebarFormContainer.js : SelectUserItem 사용
- SelectedUserItem.js : select menu에서 선택된 user item을 렌더링
- Sidebar.js : 불필요한 파일 제거

caribouuuu committed:
feat: SelectedLabelItem 컴포넌트 구현

- SelectedLabelItem.js : 선택된 label 렌더링

caribouuuu committed:
feat: SelectedMilestoneItem 컴포넌트 구현

- SelectedMilestoneItem.js : select menu에서 선택된 마일스톤 렌더링

caribouuuu committed:
feat: 새로운 issue 등록 기능 구현

- IssueFormContainer.js : onSubmit 에서 issue 데이터를 다듬어 POST /issues 요청
  - success true인 경우 새로 생성된 id 값을 통해 생성된 이슈 상세 화면으로 이동
  - success false인 경우 message alert창으로 출력

caribouuuu committed:
Merge branch 'develop' into feature/sidebar-form-fe

caribouuuu committed:
feat: image 업로드 처리를 위한 imageUploadHandler 구현

imgur로 이미지 업로드 api 요청
업로드 성공 시 전달받은 데이터에서 link(image url)을 반환

- frontend/src/config/config.js : IMGUR_CLIENT_ID 저장

caribouuuu committed:
feat: issue form 화면에서 이미지 업로드 구현

- IssueFormContainer.js : renderImageTag 함수에서 imageUploadHandler를 사용하여 이미지 태그를 이슈 내용에 append함
  - multiple 파일 업로드 구현 예정
  - '![Uploading 파일명...]()' 먼저 출력한 후 업로드 성공 시 완성된 이미지 태그를 replace 하는 기능은 미구현 (setState 연속 호출 이슈)

caribouuuu committed:
fix: issue 작성 화면에서 user profile 로그인 유저 정보로 렌더링

- CreateIssue.js : userContext 사용하여 로그아웃 유저일 경우 home으로 리다이렉트
  - 로그인 유저인 경우 IssueFormContainer에 user 정보 전달
- UserProfileContainer : 전달받은 user 데이터에서 nickname, profile_url 사용하여 컴포넌트 렌더링

caribouuuu committed:
feat: select menu item 클릭시 체크 아이콘 표시

- frontend/src/assets/icon/CheckIcon.js : 체크 아이콘 svg 컴포넌트 추가
- SelectMenuList.js : Item 컴포넌트 렌더링 시 selected:false로 생성
- UserItem.js : CheckIcon 사용
  - selected 여부로 visible 설정 -> CheckIcon 렌더링 시 visible 값 전달

caribouuuu committed:
feat: select menu - Label item 클릭시 체크 아이콘 표시

- LabelItem.js : CheckIcon 사용
  - selected 여부로 visible 설정 -> CheckIcon 렌더링 시 visible 값 전달

YimJiYoung commented on frontend/src/containers/issue-form/IssueFormContainer.js:
```diff
@@ -39,78 +13,108 @@ const FlexRowBetween = styled.div`
+      milestone_id: milestoneId,
+    };
+
+    const response = await fetch(apiUri.issues, {
+      mode: 'cors',
+      credentials: 'include',
+      method: 'POST',
+      headers: {
+        'Content-Type': 'application/json',
+      },
+      body: JSON.stringify(newIssue),
+    });
+
+    const { success, content, message } = await response.json();
+    if (!success) {
+      alert(message);
+      return;
+    }
```
이 부분도 모듈화를 해보면 좋겠다는 생각이 드네용

YimJiYoung commented on frontend/src/containers/sidebar-form/components/SelectedMilestoneItem.js:
```diff
@@ -0,0 +1,45 @@
+`;
+
+const SelectedMilestoneItem = ({ milestone }) => {
+  const {
+    id, title, description, due_date: dueDate, open_issues: openIssues, closed_issues: closedIssues, progress,
+  } = milestone;
+
+  const progressPercent = (closedIssues / (+openIssues + +closedIssues)) * 100;
```
progressPercentage 계산하는 부분은 서버에서 받아온 progress 사용하셔도 좋을 것 같습니다 .. (소곤소곤)

caribouuuu replied:
떡하니 progress 가져와놓고 다시 계산한게 제가봐도 어이없네요ㅋㅋㅋㅋㅋㅋㅋㅋㅋ 왜그랬을까요..........?

YimJiYoung commented on frontend/src/containers/sidebar-form/components/SelectMenuList.js:
```diff
@@ -14,20 +14,46 @@ const List = styled.ul`
-      const { milestones } = content;
-      list = milestones.map((item) => <MilestoneItem milestone={item} />);
+    switch (label) {
+      case selectMenuMode.Assignees[0]:
```
`selectMenuMode.Assignees[0]` , `selectMenuMode.Labels[0]` 등 이 label은 많이 사용되는 것 같아 label만 따로  모아둔 상수 객체가 있어도 괜찮을 것 같아요

caribouuuu replied:
넵 저부분은 나중에 더 깔끔하게 리팩토링 해보겠습니다..!!

YimJiYoung reviewed (approved):
이미지 업로드까지 구현하시고 컴포넌트들도 잘 분리된 것 같아요.
코드가 너무 깔끔해서 좋네요 ! 변수명들도 의미있게 지으시려고 한 것이 보입니당
수고하셨습니다 ~👏

caribouuuu committed:
refactor: milestone progress 값 사용

openissues count와 close issues count를 통해 계산하던 방식에서 progress 값 바로 사용

caribouuuu committed:
fix: user profile 컴포넌트 전달 값 수정

이슈 상세 화면 렌더링 시 user profile 컴포넌트 사용 방식의 차이로 생긴 오류 수정
nickname, profile_url을 직접 전달해주던 방식에서 user 객체를 전달해주는 방식으로 수정

caribouuuu committed:
Merge branch 'develop' into feature/sidebar-form-fe
